base/tps/src/main/java/org/dogtagpki/server/tps/rest/TPSCertService.java

```diff
@@ -162,6 +161,58 @@ public Response findCerts(String filter, String tokenID, Integer start, Integer
 
             return createOKResponse(response);
 
+        } catch (Exception e) {
+            throw new PKIException(e.getMessage());
+        }
+    }
+
+    private Response findAllCerts(String filter, Integer start, Integer size) {
+        String method = "TPSCertService:findAllCerts: ";
+        String msg = "";
+        TPSCertCollection response = new TPSCertCollection();
+
+        logger.debug("TPSCertService.findAllCerts({}, {}, {})", filter, start, size);
+
+        TPSEngine engine = TPSEngine.getInstance();
+        try {
+            TPSSubsystem subsystem = (TPSSubsystem) engine.getSubsystem(TPSSubsystem.ID);
+            TPSCertDatabase certDatabase = subsystem.getCertDatabase();
+            Iterator<TPSCertRecord> certRecs = certDatabase.findRecords(filter).iterator();
+            TokenDatabase tokenDatabase = subsystem.getTokenDatabase();
+            List<String> authorizedProfiles = getAuthorizedProfiles();
+            if (authorizedProfiles == null) {
+                msg = "authorizedProfiles null";
+                logger.debug("{}{}", method, msg);
+                throw new PKIException(method + msg);
+            }
+            if (!authorizedProfiles.contains(UserResource.ALL_PROFILES)) {
+                throw new PKIException(method + "Token records restricted");
+            }
+            int i = 0;
+
+            // return entries up to the page size
+            for (; i < start + size && certRecs.hasNext(); i++) {
+                // skip to the start of the page
+                if (i < start) certRecs.next();
```

Comment:
We should skip all the elements before the `i` start index but we are skipping only one. I do not think it is correct.

Reply:
It's probably missing a `continue` statement, but I think it will still have the same problem.

Let's use a concrete example. Suppose there are `100` certs in the database indexed from `0` to `99`, but the user only has access to the ones with even indexes (i.e. `50` certs). Suppose the user entered a keyword that matches all certs, and assuming the page size is `20`, ideally in the UI the user should see `3` pages total, each containing `20` certs except the last one containing only `10` certs.

With this code, the `certRecs` will return `100` certs because the filter will match all certs. For the first page, the loop will count `i` from `0` to `19` (i.e. `20` certs), but only add `10` certs that have the even indexes. In the second loop the `i` will be incremented until the end of the list, reaching `99`. So in the UI the user will see `5` pages total and all pages will contain only `10` certs instead of `20`.